### PR TITLE
Container specific extra environment variables

### PR DESCRIPTION
**Pull request by rishkarajgi**
---
Adds the ability to inject additional environment variables to specific containers. This currently supports addition to scheduler, webserver, worker, triggerer, flower, cleanup jobs and wait-for-migrations containers

**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by aliotta** [comment]
this PR will need chart tests

**Comment by jedcunningham** [comment]
Approach looks good. We should also do cleanup and the jobs.

File: chart/templates/dag-processor/dag-processor-deployment.yaml [hunk 1]
```diff
@@ -130,6 +130,9 @@ spec:
         {{- if .Values.dagProcessor.extraInitContainers }}
         {{- toYaml .Values.dagProcessor.extraInitContainers | nindent 8 }}
         {{- end }}
+{{- if .Values.dagProcessor.waitForMigrations.extraEnv }}
+{{ tpl (toYaml .Values.dagProcessor.waitForMigrations.extraEnv) $ | indent 12 }}
```
Comment (jedcunningham) [inline chart/templates/dag-processor/dag-processor-deployment.yaml]: Only waitForMigrations? What about the normal container?

File: chart/templates/workers/worker-deployment.yaml [hunk 2]
```diff
@@ -219,6 +219,9 @@ spec:
         {{- if and (.Values.dags.gitSync.enabled) (not .Values.dags.persistence.enabled) }}
         {{- include "git_sync_container" . | indent 8 }}
         {{- end }}
+{{- if .Values.workers.extraEnv }}
+{{ tpl (toYaml .Values.workers.extraEnv) $ | indent 12 }}
```
Comment (jedcunningham) [inline chart/templates/workers/worker-deployment.yaml]: Why no waitForMigrations here? Don't forget the k8s pod_template_file too.

File: chart/values.schema.json [hunk 3]
```diff
@@ -552,6 +552,11 @@
                                 "IfNotPresent"
                             ],
                             "default": "IfNotPresent"
+                        },
+                        "extraEnv": {
+                            "description": "Add additional env vars to flower.",
+                            "type": "array",
+                            "default": []
```
Comment (jedcunningham) [inline chart/values.schema.json]: You should add "items" here so we can properly validate what the user supplies here. e.g:
https://github.com/apache/airflow/blob/2af19f16a4d94e749bbf6c7c4704e02aac35fc11/chart/values.schema.json#L695-L710

**Review by jedcunningham — CHANGES_REQUESTED** [review]
You should use `env`, not `extraEnv`. `extraEnv` is already used globally for thing based config, so having one a string and another a list is confusing:
 
https://github.com/apache/airflow/blob/2af19f16a4d94e749bbf6c7c4704e02aac35fc11/chart/values.yaml#L300-L307

File: chart/values.schema.json (hunk 3 above)
Comment (rishkarajgi) [reply chart/values.schema.json]: Have added "items" as standard validation for envVar
```
"items": {
"$ref": "#/definitions/io.k8s.api.core.v1.EnvVar"
}
```

File: chart/templates/dag-processor/dag-processor-deployment.yaml (hunk 1 above)
Comment (rishkarajgi) [reply chart/templates/dag-processor/dag-processor-deployment.yaml]: Wasn't aware if the dag-processor container interacted with the airflow db. Have added it for the main container as well

**Comment by rishkarajgi** [comment]
> You should use `env`, not `extraEnv`. `extraEnv` is already used globally for thing based config, so having one a string and another a list is confusing:
> 
> https://github.com/apache/airflow/blob/2af19f16a4d94e749bbf6c7c4704e02aac35fc11/chart/values.yaml#L300-L307

@jedcunningham Have updated `extraEnv` to `env` as a property under each container

File: chart/templates/workers/worker-deployment.yaml (hunk 2 above)
Comment (rishkarajgi) [reply chart/templates/workers/worker-deployment.yaml]: @jedcunningham There isn't currently a waitForMigrations property for worker pods. I have added the same set of envs used for the worker container for now. Or I could introduce a new property to add custom envs for the migrations container.
Could you direct me on how to overwrite the base pod_template_file with labels and envs?

Comment (jedcunningham) [reply chart/templates/workers/worker-deployment.yaml]: Here is the pod_template_file: https://github.com/apache/airflow/blob/main/chart/files/pod-template-file.kubernetes-helm-yaml

File: chart/Chart.yaml [hunk 4]
```diff
@@ -19,7 +19,7 @@
 ---
 apiVersion: v2
 name: airflow
-version: 1.7.0-dev
+version: 1.8.0-dev
```
Comment (jedcunningham) [inline chart/Chart.yaml]:
```suggestion
version: 1.7.0-dev
```

This gets bumped as part of the chart release process.

Comment (jedcunningham) [inline chart/Chart.yaml]:
```suggestion
version: 1.7.0-dev
```

This gets bumped as part of the release process.

**Review by jedcunningham — COMMENTED** [review]
Let's keep this PR to just env vars. Labels should be done in a separate PR.

File: chart/files/pod-template-file.kubernetes-helm-yaml [hunk 5]
```diff
@@ -58,6 +58,9 @@ spec:
           value: LocalExecutor
 {{- include "standard_airflow_environment" . | indent 6}}
 {{- include "custom_airflow_environment" . | indent 6 }}
+{{- if .Values.workers.env }}
+{{ tpl (toYaml .Values.workers.env) $ | indent 12 }}
+{{- end }}
```
Comment (rishkarajgi) [inline chart/files/pod-template-file.kubernetes-helm-yaml]: @jedcunningham have made the changes to the pod template file here

File: chart/templates/workers/worker-deployment.yaml (hunk 2 above)
Comment (rishkarajgi) [reply chart/templates/workers/worker-deployment.yaml]: Thanks. Have pushed changes to the pod template file

**Comment by rishkarajgi** [comment]
> Let's keep this PR to just env vars. Labels should be done in a separate PR.

Had to merge the two PRs to generate a helm chart for testing. Shall keep it in two separate PRs

File: chart/templates/dag-processor/dag-processor-deployment.yaml [hunk 6]
```diff
@@ -130,6 +130,9 @@ spec:
         {{- if .Values.dagProcessor.extraInitContainers }}
         {{- toYaml .Values.dagProcessor.extraInitContainers | nindent 8 }}
         {{- end }}
+{{- if .Values.dagProcessor.waitForMigrations.env }}
+{{ tpl (toYaml .Values.dagProcessor.waitForMigrations.env) $ | indent 12 }}
+{{- end }}
       containers:
         - name: dag-processor
```
Comment (jedcunningham) [inline chart/templates/dag-processor/dag-processor-deployment.yaml]: You've missed adding `dagProcessor.env` to the main container here.

Comment (rishkarajgi) [reply chart/templates/dag-processor/dag-processor-deployment.yaml]: Ah thanks! Added the snippet

File: chart/values.schema.json [hunk 7]
```diff
@@ -2619,6 +2698,14 @@
                     "description": "Specify if you want to use the default Helm Hook annotations",
                     "type": "boolean",
                     "default": true
+                },
+                "env": {
+                    "description": "Add additional env vars to job.",
```
Comment (jedcunningham) [inline chart/values.schema.json]:
```suggestion
                    "description": "Add additional env vars to the create user job pod.",
```

File: chart/values.yaml [hunk 8]
```diff
@@ -597,6 +597,12 @@ workers:
     #   cpu: 100m
     #   memory: 128Mi
 
+  waitForMigrations:
+    # To add env vars specific to wait-for-migrations init container
```
Comment (jedcunningham) [inline chart/values.yaml]:
```suggestion
```

File: tests/charts/test_pod_template_file.py [hunk 9]
```diff
@@ -639,6 +639,16 @@ def test_should_add_pod_labels(self):
             "tier": "airflow",
         } == jmespath.search("metadata.labels", docs[0])
 
+    def test_should_add_extraEnvs(self):
+        docs = render_chart(
+            values={"env": [{"name": "TEST_ENV_1", "value": "test_env_1"}]},
```
Comment (jedcunningham) [inline tests/charts/test_pod_template_file.py]:
```suggestion
            values={"worker": {"env": [{"name": "TEST_ENV_1", "value": "test_env_1"}]}},
```

File: tests/charts/test_scheduler.py [hunk 10]
```diff
@@ -115,6 +115,39 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
             "spec.template.spec.containers[0].volumeMounts[*].name", docs[0]
         )
 
+    def test_should_add_extraEnvs(self):
+        docs = render_chart(
+            values={
+                "executor": "CeleryExecutor",
```
Comment (jedcunningham) [inline tests/charts/test_scheduler.py]:
```suggestion
```

This is the default.

File: tests/charts/test_scheduler.py [hunk 11]
```diff
@@ -115,6 +115,39 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
...
+        )
+
+        assert {'name': 'TEST_ENV_1', 'value': 'test_env_1'} in jmespath.search(
+            "spec.template.spec.containers[0].env", docs[0]
+        )
+
+    def test_should_add_extraEnvs_to_wait_for_migration_container(self):
+        docs = render_chart(
+            values={
+                "executor": "CeleryExecutor",
```
Comment (jedcunningham) [inline tests/charts/test_scheduler.py]:
```suggestion
```

File: tests/charts/test_scheduler.py [hunk 12]
```diff
@@ -115,6 +115,39 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
...
+                "scheduler": {
+                    "waitForMigrations": {
+                        "enabled": True,
```
Comment (jedcunningham) [inline tests/charts/test_scheduler.py]:
```suggestion
```

Also the default.

File: tests/charts/test_triggerer.py [hunk 13]
```diff
@@ -131,6 +131,39 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
             "spec.template.spec.containers[0].volumeMounts[0].name", docs[0]
         )
 
+    def test_should_add_extraEnvs(self):
+        docs = render_chart(
+            values={
+                "triggerer": {
+                    "enabled": True,
```
Comment (jedcunningham) [inline tests/charts/test_triggerer.py]:
```suggestion
```

File: tests/charts/test_triggerer.py [hunk 14]
```diff
@@ -131,6 +131,39 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
...
+            "spec.template.spec.containers[0].env", docs[0]
+        )
+
+    def test_should_add_extraEnvs_to_wait_for_migration_container(self):
+        docs = render_chart(
+            values={
+                "triggerer": {
+                    "enabled": True,
+                    "waitForMigrations": {
+                        "enabled": True,
```
Comment (jedcunningham) [inline tests/charts/test_triggerer.py]:
```suggestion
                    "waitForMigrations": {
```

File: tests/charts/test_webserver.py [hunk 15]
```diff
@@ -169,6 +169,39 @@ def test_should_add_extra_containers(self):
             "image": "test-registry/test-repo:test-tag",
         } == jmespath.search("spec.template.spec.containers[-1]", docs[0])
 
+    def test_should_add_extraEnvs(self):
+        docs = render_chart(
+            values={
+                "executor": "CeleryExecutor",
```
Comment (jedcunningham) [inline tests/charts/test_webserver.py]:
```suggestion
```

File: tests/charts/test_webserver.py [hunk 16]
```diff
@@ -169,6 +169,39 @@ def test_should_add_extra_containers(self):
...
+            "spec.template.spec.containers[0].env", docs[0]
+        )
+
+    def test_should_add_extraEnvs_to_wait_for_migration_container(self):
+        docs = render_chart(
+            values={
+                "executor": "CeleryExecutor",
+                "webserver": {
+                    "waitForMigrations": {
+                        "enabled": True,
```
Comment (jedcunningham) [inline tests/charts/test_webserver.py]:
```suggestion
                "webserver": {
                    "waitForMigrations": {
```

File: tests/charts/test_worker.py [hunk 17]
```diff
@@ -112,6 +112,36 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
             "spec.template.spec.containers[0].volumeMounts[0].name", docs[0]
         )
 
+    def test_should_add_extraEnvs(self):
+        docs = render_chart(
+            values={
+                "executor": "CeleryExecutor",
```
Comment (jedcunningham) [inline tests/charts/test_worker.py]:
```suggestion
```

File: tests/charts/test_worker.py [hunk 18]
```diff
@@ -112,6 +112,36 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
...
+        )
+
+        assert {'name': 'TEST_ENV_1', 'value': 'test_env_1'} in jmespath.search(
+            "spec.template.spec.containers[0].env", docs[0]
+        )
+
+    def test_should_add_extraEnvs_to_wait_for_migration_container(self):
+        docs = render_chart(
+            values={
+                "executor": "CeleryExecutor",
```
Comment (jedcunningham) [inline tests/charts/test_worker.py]:
```suggestion
```

File: chart/templates/dag-processor/dag-processor-deployment.yaml [hunk 19]
```diff
@@ -134,6 +134,9 @@ spec:
         {{- if .Values.dagProcessor.extraInitContainers }}
         {{- toYaml .Values.dagProcessor.extraInitContainers | nindent 8 }}
         {{- end }}
+{{- if .Values.dagProcessor.waitForMigrations.env }}
+{{ tpl (toYaml .Values.dagProcessor.waitForMigrations.env) $ | indent 12 }}
+{{- end }}
```
Comment (jedcunningham) [inline chart/templates/dag-processor/dag-processor-deployment.yaml]: This isn't in the right spot. Needs to move up to ensure it's in the envs.

File: chart/templates/workers/worker-deployment.yaml [hunk 20]
```diff
@@ -223,6 +226,9 @@ spec:
         {{- if and (.Values.dags.gitSync.enabled) (not .Values.dags.persistence.enabled) }}
         {{- include "git_sync_container" . | indent 8 }}
         {{- end }}
+{{- if .Values.workers.env }}
+{{ tpl (toYaml .Values.workers.env) $ | indent 12 }}
+{{- end }}
```
Comment (jedcunningham) [inline chart/templates/workers/worker-deployment.yaml]: Same here, needs to be moved to the right spot.

File: chart/files/pod-template-file.kubernetes-helm-yaml [hunk 21]
```diff
@@ -58,6 +58,9 @@ spec:
           value: LocalExecutor
 {{- include "standard_airflow_environment" . | indent 6}}
 {{- include "custom_airflow_environment" . | indent 6 }}
+{{- if .Values.workers.env }}
+{{ tpl (toYaml .Values.workers.env) $ | indent 12 }}
```
Comment (jedcunningham) [inline chart/files/pod-template-file.kubernetes-helm-yaml]:
```suggestion
{{ tpl (toYaml .Values.workers.env) $ | indent 8 }}
```

I think this is the right indent?

**Review by jedcunningham — CHANGES_REQUESTED** [review]

File: chart/values.schema.json (hunk 7 above)
Comment (rishkarajgi) [reply chart/values.schema.json]: done

File: chart/values.yaml (hunk 8 above)
Comment (rishkarajgi) [reply chart/values.yaml]: done

File: tests/charts/test_scheduler.py (hunk 10 above)
Comment (rishkarajgi) [reply tests/charts/test_scheduler.py]: done

File: tests/charts/test_scheduler.py (hunk 11 above)
Comment (rishkarajgi) [reply tests/charts/test_scheduler.py]: done

File: tests/charts/test_scheduler.py (hunk 12 above)
Comment (rishkarajgi) [reply tests/charts/test_scheduler.py]: done

File: tests/charts/test_triggerer.py (hunk 13 above)
Comment (rishkarajgi) [reply tests/charts/test_triggerer.py]: done

File: tests/charts/test_triggerer.py (hunk 14 above)
Comment (rishkarajgi) [reply tests/charts/test_triggerer.py]: done

File: tests/charts/test_webserver.py (hunk 15 above)
Comment (rishkarajgi) [reply tests/charts/test_webserver.py]: done

File: tests/charts/test_webserver.py (hunk 16 above)
Comment (rishkarajgi) [reply tests/charts/test_webserver.py]: done

File: tests/charts/test_worker.py (hunk 17 above)
Comment (rishkarajgi) [reply tests/charts/test_worker.py]: done

File: tests/charts/test_worker.py (hunk 18 above)
Comment (rishkarajgi) [reply tests/charts/test_worker.py]: done

File: chart/templates/dag-processor/dag-processor-deployment.yaml (hunk 19 above)
Comment (rishkarajgi) [reply chart/templates/dag-processor/dag-processor-deployment.yaml]: @jedcunningham I ran a helm template and it seems to be aligning right under the env section

Comment (jedcunningham) [reply chart/templates/dag-processor/dag-processor-deployment.yaml]: Try it with gitsync enabled or an extra init container.

File: chart/files/pod-template-file.kubernetes-helm-yaml (hunk 21 above)
Comment (rishkarajgi) [reply chart/files/pod-template-file.kubernetes-helm-yaml]: done

File: chart/templates/workers/worker-deployment.yaml (hunk 20 above)
Comment (rishkarajgi) [reply chart/templates/workers/worker-deployment.yaml]: done

File: chart/templates/dag-processor/dag-processor-deployment.yaml (hunk 19 above)
Comment (rishkarajgi) [reply chart/templates/dag-processor/dag-processor-deployment.yaml]: done

File: tests/charts/test_pod_template_file.py (hunk 9 above)
Comment (rishkarajgi) [reply tests/charts/test_pod_template_file.py]: done

File: chart/templates/_helpers.yaml [hunk 22]
```diff
@@ -782,3 +782,83 @@ Where `.` is the global variables scope and `.Values.workers` the local variable
     {{- end -}}
   {{- end -}}
 {{- end -}}
+{{- define "workers_extra_envs" -}}
```
Comment (jedcunningham) [inline chart/templates/_helpers.yaml]: You shouldn't need to do a helper per component, just pass the right map in from the template and have a single helper.

Comment (rishkarajgi) [reply chart/templates/_helpers.yaml]: Done

File: chart/values.schema.json [hunk 23]
```diff
@@ -561,6 +561,27 @@
...
+                                    "value": {
+                                        "type": "string"
+                                    }
+                                },
+                                "required": [
+                                    "name",
+                                    "value"
+                                ],
+                                "additionalProperties": false
+                            }
```
Comment (jedcunningham) [inline chart/values.schema.json]:
```suggestion
```

Don't need this, wrong section.

File: tests/charts/test_dag_processor.py [hunk 24]
```diff
@@ -119,6 +119,39 @@ def test_should_add_extra_volume_and_extra_volume_mount(self):
...
+            "spec.template.spec.containers[0].env", docs[0]
+        )
+
+    def test_should_add_extraEnvs_to_wait_for_migration_container(self):
+        docs = render_chart(
+            values={
+                "dagProcessor": {
+                    "enabled": True,
+                    "waitForMigrations": {
+                        "enabled": True,
```
Comment (jedcunningham) [inline tests/charts/test_dag_processor.py]:
```suggestion
```

This is the default.

Comment (rishkarajgi) [reply tests/charts/test_dag_processor.py]: resolved

File: chart/values.schema.json (hunk 23 above)
Comment (rishkarajgi) [reply chart/values.schema.json]: done

**Review by jedcunningham — APPROVED** [review]

**Pull request merged**